File: CS/XtraReport2.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraReports.UI;
using DevExpress.DataAccess.Native.Sql;
using DevExpress.DataAccess.Sql;
using DevExpress.DataAccess.Sql.DataApi;

namespace dxSample {
    public partial class XtraReport2 : DevExpress.XtraReports.UI.XtraReport {
        public XtraReport2() {
            InitializeComponent();
        }
        string FilterExpression{
            get {
                return String.Format("CategoryName == '{0}'", this.Parameters["categoryName"].Value);
            }
        }
        private void Detail_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            e.Cancel = !IsFit();
        }

        private bool IsFit() {
            ResultRow rr = this.GetCurrentRow() as ResultRow;

            IResultSet resultSet = (this.DataSource as SqlDataSource).Result as DevExpress.DataAccess.Sql.DataApi.IResultSet;
            ITable tableProd = null;
            resultSet.TryGetTable("CategoryProducts", out tableProd);
            DevExpress.Data.Filtering.CriteriaOperator filter = DevExpress.Data.Filtering.CriteriaOperator.Parse(FilterExpression);
            DevExpress.Data.Filtering.Helpers.ExpressionEvaluator evaluator = new DevExpress.Data.Filtering.Helpers.ExpressionEvaluator((((ITypedList)tableProd).GetItemProperties(null)), filter);
            bool isFit = evaluator.Fit(rr);
            return isFit;
        }

        private void GroupHeader1_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            e.Cancel = !IsFit();
        }
    }
}

File: VB/XtraReport2.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.XtraReports.UI
Imports DevExpress.DataAccess.Native.Sql
Imports DevExpress.DataAccess.Sql
Imports DevExpress.DataAccess.Sql.DataApi

Namespace dxSample
    Partial Public Class XtraReport2
        Inherits DevExpress.XtraReports.UI.XtraReport

        Public Sub New()
            InitializeComponent()
        End Sub
        Private ReadOnly Property FilterExpression() As String
            Get
                Return String.Format("CategoryName == '{0}'", Me.Parameters("categoryName").Value)
            End Get
        End Property
        Private Sub Detail_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles Detail.BeforePrint
            e.Cancel = Not IsFit()
        End Sub

        Private Function IsFit() As Boolean
            Dim rr As ResultRow = TryCast(Me.GetCurrentRow(), ResultRow)

            Dim resultSet As IResultSet = TryCast((TryCast(Me.DataSource, SqlDataSource)).Result, DevExpress.DataAccess.Sql.DataApi.IResultSet)
            Dim tableProd As ITable = Nothing
            resultSet.TryGetTable("CategoryProducts", tableProd)
            Dim filter As DevExpress.Data.Filtering.CriteriaOperator = DevExpress.Data.Filtering.CriteriaOperator.Parse(FilterExpression)
            Dim evaluator As New DevExpress.Data.Filtering.Helpers.ExpressionEvaluator((DirectCast(tableProd, ITypedList).GetItemProperties(Nothing)), filter)

            Dim isFit_Renamed As Boolean = evaluator.Fit(rr)
            Return isFit_Renamed
        End Function

        Private Sub GroupHeader1_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles GroupHeader1.BeforePrint
            e.Cancel = Not IsFit()
        End Sub
    End Class
End Namespace
